File: eMRTDReader/sdk/src/main/java/com/example/emrtdreader/data/NfcPassportReader.java
package com.example.emrtdreader.data;

import android.content.Context;
import android.nfc.Tag;
import android.nfc.tech.IsoDep;

import com.example.emrtdreader.crypto.PassiveAuthVerifier;
import com.example.emrtdreader.domain.AccessKey;
import com.example.emrtdreader.domain.PassportReadResult;
import com.example.emrtdreader.error.PassportReadException;
import com.example.emrtdreader.models.PassportChipData;
import com.example.emrtdreader.models.VerificationResult;

import net.sf.scuba.smartcards.IsoDepCardService;

import org.jmrtd.PassportService;
import org.jmrtd.lds.icao.DG1File;
import org.jmrtd.lds.icao.DG2File;

import java.io.ByteArrayInputStream;
import java.util.HashMap;
import java.util.Map;

public class NfcPassportReader {
    private static final int ISO_DEP_TIMEOUT_MS = 12000;
    private static final int MAX_RETRIES = 2;

    public PassportReadResult read(Context ctx, Tag tag, AccessKey.Mrz mrz, String can) throws PassportReadException {
        IsoDep isoDep = IsoDep.get(tag);
        if (isoDep == null) throw new PassportReadException.TagNotIsoDep();

        isoDep.setTimeout(ISO_DEP_TIMEOUT_MS);

        PassportReadException last = null;

        for (int attempt=0; attempt<=MAX_RETRIES; attempt++) {
            IsoDepCardService cardService = new IsoDepCardService(isoDep);
            PassportService ps = null;

            try {
                cardService.open();

                ps = new PassportService(
                        cardService,
                        PassportService.NORMAL_MAX_TRANCEIVE_LENGTH,
                        PassportService.DEFAULT_MAX_BLOCKSIZE,
                        false,
                        false
                );
                ps.open();

                try { ps.sendSelectApplet(false); } catch (Throwable ignore) {}

                new JmrtdAccessController(ps).establish(mrz, can);

                byte[] dg1Bytes = readEfOrThrow(ps, PassportService.EF_DG1);
                byte[] dg2Bytes = readEfOrNull(ps, PassportService.EF_DG2);
                byte[] sodBytes = readEfOrNull(ps, PassportService.EF_SOD);

                DG1File dg1 = new DG1File(new ByteArrayInputStream(dg1Bytes));
                var mrzInfo = dg1.getMRZInfo();

                byte[] photo = null;
                if (dg2Bytes != null) {
                    try {
                        DG2File dg2 = new DG2File(new ByteArrayInputStream(dg2Bytes));
                        if (!dg2.getFaceInfos().isEmpty()
                                && !dg2.getFaceInfos().get(0).getFaceImageInfos().isEmpty()) {
                            photo = dg2.getFaceInfos().get(0).getFaceImageInfos().get(0).getImageInputStream().readAllBytes();
                        }
                    } catch (Throwable ignore) {}
                }

                Map<Integer, byte[]> dgs = new HashMap<>();
                dgs.put(1, dg1Bytes);
                if (dg2Bytes != null) dgs.put(2, dg2Bytes);

                VerificationResult vr = PassiveAuthVerifier.verify(ctx, sodBytes, dgs);

                PassportChipData chip = new PassportChipData(
                        mrzInfo.getDocumentNumber(),
                        mrzInfo.getPrimaryIdentifier(),
                        mrzInfo.getSecondaryIdentifier(),
                        mrzInfo.getNationality(),
                        mrzInfo.getDateOfBirth(),
                        String.valueOf(mrzInfo.getGender()),
                        mrzInfo.getDateOfExpiry(),
                        photo
                );

                return new PassportReadResult(chip, vr);

            } catch (PassportReadException e) {
                last = e;
            } catch (Throwable t) {
                last = new PassportReadException.ReadFailed(t);
            } finally {
                try { if (ps != null) ps.close(); } catch (Throwable ignore) {}
                try { cardService.close(); } catch (Throwable ignore) {}
                // do NOT close isoDep here; Android will manage tag lifecycle; but we can try:
                // try { isoDep.close(); } catch (Throwable ignore) {}
            }

            // retry on TagLost/IO-like errors
            if (last != null && attempt < MAX_RETRIES) {
                try { Thread.sleep(150); } catch (InterruptedException ignore) {}
                continue;
            }
        }

        throw last != null ? last : new PassportReadException.ReadFailed(new IllegalStateException("Unknown NFC error"));
    }

    private byte[] readEfOrThrow(PassportService ps, short fid) throws Exception {
        byte[] b = readEfOrNull(ps, fid);
        if (b == null) throw new IllegalStateException("Required EF missing: " + fid);
        return b;
    }

    private byte[] readEfOrNull(PassportService ps, short fid) {
        try {
            return ps.getInputStream(fid).readAllBytes();
        } catch (Throwable t) {
            return null;
        }
    }
}

File: eMRTDReader/app/src/main/java/com/example/emrtdreader/data/NfcPassportReader.kt
package com.example.emrtdreader.data

import android.nfc.Tag
import android.nfc.tech.IsoDep
import com.example.emrtdreader.crypto.AuthResult
import com.example.emrtdreader.crypto.CryptoHelper
import com.example.emrtdreader.domain.PassportReadResult
import com.example.emrtdreader.error.PassportReadException
import com.example.emrtdreader.models.AccessKey
import com.example.emrtdreader.models.MrzResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.sf.scuba.smartcards.CardService
import net.sf.scuba.smartcards.IsoDepCardService
import org.jmrtd.BACKey
import org.jmrtd.PassportService
import org.jmrtd.lds.icao.DG1File
import org.jmrtd.lds.icao.DG2File
import org.jmrtd.lds.SODFile
import java.io.ByteArrayInputStream

class NfcPassportReader {

    suspend fun read(tag: Tag, mrzResult: MrzResult): PassportReadResult = withContext(Dispatchers.IO) {
        val isoDep = IsoDep.get(tag) ?: throw PassportReadException.TagNotIsoDep()
        val cardService = IsoDepCardService(isoDep)

        try {
            cardService.open()
            val passportService = PassportService(cardService, PassportService.NORMAL_MAX_TRANCEIVE_LENGTH, PassportService.DEFAULT_MAX_BLOCKSIZE, false, false)
            passportService.open()

            // Extract info from the (already validated) MRZ Result
            val docNum = mrzResult.line2.substring(0, 9)
            val dob = mrzResult.line2.substring(13, 19)
            val doe = mrzResult.line2.substring(21, 27)

            val bacKey = BACKey(docNum, dob, doe)
            passportService.sendSelectApplet(false)
            passportService.doBAC(bacKey)

            val dg1Stream = passportService.getInputStream(PassportService.EF_DG1)
            val dg1Bytes = dg1Stream.readBytes()
            val dg1 = DG1File(ByteArrayInputStream(dg1Bytes))

            val dg2Stream = try { passportService.getInputStream(PassportService.EF_DG2) } catch (e: Exception) { null }
            val sodStream = try { passportService.getInputStream(PassportService.EF_SOD) } catch (e: Exception) { null }

            val sodBytes = sodStream?.readBytes()
            val dg2Bytes = dg2Stream?.readBytes()

            val authResult = if (sodBytes != null) {
                val sodFile = SODFile(ByteArrayInputStream(sodBytes))
                val dataHashes = mutableMapOf<Int, ByteArray>()
                dataHashes[1] = dg1Bytes
                dg2Bytes?.let { dataHashes[2] = it }
                
                CryptoHelper.verifyPassiveAuth(sodFile, dataHashes)
            } else {
                AuthResult.SOD_NOT_FOUND
            }

            val photo = dg2Bytes?.let { DG2File(ByteArrayInputStream(it)).faceInfos.firstOrNull()?.faceImageInfos?.firstOrNull()?.imageInputStream?.readBytes() }

            val passportData = PassportData(
                mrz = mrzResult.line1 + "\n" + mrzResult.line2 + (mrzResult.line3?.let { "\n" + it } ?: ""),
                dg1 = dg1Bytes,
                dg2 = dg2Bytes,
                sod = sodBytes,
                photo = photo
            )

            PassportReadResult(passportData, authResult, "")

        } catch (e: Exception) {
            throw PassportReadException.ReadFailed(e)
        } finally {
            cardService.close()
        }
    }
}